android/src/main/java/com/pauldemarco/flutter_blue/FlutterBluePlugin.java

```diff
@@ -364,128 +375,166 @@ public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
                 break;
             }
 
-            case "isAvailable":
-            {
+            case "isAvailable": {
                 result.success(mBluetoothAdapter != null);
                 break;
             }
 
-            case "isOn":
-            {
+            case "isOn": {
                 result.success(mBluetoothAdapter.isEnabled());
                 break;
             }
 
-            case "startScan":
-            {
-                scanForBLE(Pattern.compile(".*GoPro.*"));
-                result.success(null);
+            case "startScan": {
+                ensurePermissionBeforeAction(Manifest.permission.BLUETOOTH_CONNECT, null, new OperationOnPermission() {
+                    @Override
+                    public void op(boolean granted, String permission) {
+                        if (!granted) {
+                            FlutterBluePlugin.this.log(LogLevel.ERROR, "No permission for bluetooth while start scanning ");
+
+                            result.error(
+                                    "no_permissions", String.format("flutter_blue plugin requires %s for scanning", permission), null);
+
+                        } else {
+                            FlutterBluePlugin.this.scanForBLE(Pattern.compile(".*GoPro.*"));
+                            result.success(null);
+                        }
+                    }
+                });
                 break;
             }
 
-            case "stopScan":
-            {
+            case "stopScan": {
                 // Nothing to do as we don't do explicit scans
                 result.success(null);
                 break;
             }
 
-            case "getConnectedDevices":
-            {
-                List<BluetoothDevice> devices = mBluetoothManager.getConnectedDevices(BluetoothProfile.GATT);
-                Protos.ConnectedDevicesResponse.Builder p = Protos.ConnectedDevicesResponse.newBuilder();
-                for(BluetoothDevice d : devices) {
-                    p.addDevices(ProtoMaker.from(d));
-                }
-                result.success(p.build().toByteArray());
-                log(LogLevel.EMERGENCY, "mDevices size: " + mDevices.size());
+            case "getConnectedDevices": {
+                ensurePermissionBeforeAction(Manifest.permission.BLUETOOTH_CONNECT, null, new OperationOnPermission() {
+                    @Override
+                    public void op(boolean granted, String permission) {
+                        if (!granted) {
+                            FlutterBluePlugin.this.log(LogLevel.ERROR, "No permission for bluetooth while getting connected devices ");
```

Comment:
This code can be made into a function. It's almost the same for all cases? Can this be made into something like this
```
checkForBluetoothConnect() {
  @Override
  public void onSuccess() {
    // Do the positive case
  }
}
```

Reply:
We have to send `result.error` in method channel if any error occur so we need to maintain both `result.success` and `result.error`. I think current function is more dynamic as compared to `checkForBluetoothConnect` because if tomorrow another permission change for any case so we just add/subtract for particular case.